e2e_test/batch/functions/array_concat.slt.part

```diff
@@ -0,0 +1,124 @@
+query T
+select array_cat(array[66], array[123]);
+----
+{66,123}
+
+query T
+select array_cat(array[66], null::int[]);
+----
+{66}
+
+query T
+select array_cat(null::int[], array[123]);
+----
+{123}
+
+query T
+select array_cat(array[array[66]], array[233]);
+----
+{{66},{233}}
+
+query T
+select array_cat(array[array[66]], null::int[]);
+----
+{{66}}
+
+query T
+select array_cat(null::int[][], array[233]);
```

Comment:
> ```sql
> select array_cat(array[123], null::int[][]);
> ```
>
> PG returns `array[123]` for the above SQL, however, we save element type in `List` type and have to infer the return type in frontend, hence can't return `array[123]` once we decide it's `array_cat(int[], int[][])`.

Would returning `array[array[123]]` a better choice?

IMO PG's behavior is quite strange, especially in this example:
```sql
create table t(x int[], y int[][]);
select x, y, array_cat(x,y) from t;
  x  |   y   | array_cat
-----+-------+-----------
 {1} | {{2}} | {{1},{2}}
 {1} |       | {1}
(2 rows)
```

(Off topic) Other issues about PG array https://github.com/risingwavelabs/risingwave/issues/3811

Reply:
(mentioned by @richardchien off-line): PG do not allow multi dimentional array with mismatching dimensions, this is related to #3811.

For `array_cat`:
```sql
select array_cat(Array[1], Array[Array[2,3]]);
ERROR:  cannot concatenate incompatible arrays
DETAIL:  Arrays with differing dimensions are not compatible for concatenation.

select array_cat(Array[Array[1]], Array[Array[2,3]]);
ERROR:  cannot concatenate incompatible arrays
DETAIL:  Arrays with differing element dimensions are not compatible for concatenation.
```

> we do lack a clear definition/expectation of what we support.

😇

BTW, maybe add such cases to the test cases to show our current behavior?

Reply:
Things can be more funny: 😄
```sql
select x, y, array_cat(x,y) from t;
    x    |   y   |   array_cat
---------+-------+---------------
 {1}     | {{2}} | {{1},{2}}
 {1}     |       | {1}
 {{1}}   | {{1}} | {{1},{1}}
 {{{1}}} | {{1}} | {{{1}},{{1}}}
(4 rows)
```